src/storage/src/hummock/file_cache/store.rs

```diff
@@ -0,0 +1,221 @@
+
+        // Get file system type and block size by `statfs(2)`.
+        let fs_stat = statfs(options.dir.as_str())?;
+        let fs_type = match fs_stat.filesystem_type() {
+            EXT4_SUPER_MAGIC => FsType::Ext4,
+            // FYI: https://github.com/nix-rust/nix/issues/1742
+            NixFsType(libc::XFS_SUPER_MAGIC) => FsType::Xfs,
+            nix_fs_type => return Err(Error::UnsupportedFilesystem(nix_fs_type.0)),
+        };
+        let fs_block_size = fs_stat.block_size() as usize;
+
+        let cf_opts = CacheFileOptions {
+            fs_block_size,
+            // TODO: Make it configurable.
+            block_size: fs_block_size,
+            fallocate_unit: options.cache_file_fallocate_unit,
+        };
+
+        let mf = MetaFile::open(PathBuf::from(&options.dir).join(META_FILE_FILENAME))?;
+
+        let cf = CacheFile::open(
+            PathBuf::from(&options.dir).join(CACHE_FILE_FILENAME),
+            cf_opts,
+        )
+        .await?;
+
+        Ok(Self {
+            _dir: options.dir,
+            _capacity: options.capacity,
+
+            _fs_type: fs_type,
+            _fs_block_size: fs_block_size,
+            // TODO: Make it configurable.
+            block_size: fs_block_size,
+            buffer_capacity: options.buffer_capacity,
+
+            mf: Arc::new(RwLock::new(mf)),
+            cf,
+        })
+    }
+
+    pub fn block_size(&self) -> usize {
+        self.block_size
+    }
+
+    pub fn size(&self) -> usize {
+        self.cf.size() + self.mf.read().size()
+    }
+
+    pub fn meta_file_size(&self) -> usize {
+        self.mf.read().size()
+    }
+
+    pub fn cache_file_size(&self) -> usize {
+        self.cf.size()
+    }
+
+    pub fn cache_file_len(&self) -> usize {
+        self.cf.len()
+    }
+
+    pub async fn restore(&self, _indices: &LruCache<K, SlotId>) -> Result<()> {
+        // TODO: Impl me!!!
+        Ok(())
+    }
+
+    pub async fn insert(&self, batch: &[(K, Vec<u8>)]) -> Result<Vec<SlotId>> {
+        debug_assert!(!batch.is_empty());
+        let mut buf = DioBuffer::with_capacity_in(self.buffer_capacity, &DIO_BUFFER_ALLOCATOR);
+        let mut blocs = Vec::with_capacity(batch.len());
+        let mut slots = Vec::with_capacity(batch.len());
+
+        for (_key, value) in batch {
+            debug_assert!(!value.is_empty());
+            let bloc = BlockLoc {
+                bidx: buf.len() as u32 / self.block_size as u32,
+                len: value.len() as u32,
+            };
+            blocs.push(bloc);
+            buf.extend_from_slice(value);
+            buf.resize(utils::usize::align_up(self.block_size, buf.len()), 0);
+        }
+
+        let boff = self.cf.append(buf).await? as u32 / self.block_size as u32;
+
+        for bloc in &mut blocs {
+            bloc.bidx += boff;
+        }
+
+        let mut mf = self.mf.write();
+        for ((key, _value), bloc) in batch.iter().zip_eq(blocs.iter()) {
+            slots.push(mf.insert(key, bloc)?);
```

Comment:
For the meta file is a memory-mapped file, `mf.insert` only requires memory operations, except when growing the capacity of the meta file. So, `mf.insert` can only fail because of `mremap(2)` and `fallocate(2)`, which cannot be handled trivially, so panic is an acceptable behavior, IMO.

Reply:
Shall we panic `mf.insert` then?

Reply:
I prefer raise the error here. But panic is also okey. 🤔

Reply:
Can add some doc for this case ?

Reply:
> Can add some doc for this case ?

IMO the error message is enough to cover for now.